src/outputs/history.cpp

```diff
@@ -49,68 +49,111 @@ void HistoryOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm,
   if (signal == SignalHandler::OutputSignal::now) {
     return;
   }
-  std::vector<std::string> all_labels = {};
-  std::vector<Real> all_results = {};
+
+  std::map<UserHistoryOperation, std::vector<Real>> results;
+  std::map<UserHistoryOperation, std::vector<std::string>> labels;
+  std::vector<UserHistoryOperation> ops = {
+      UserHistoryOperation::sum, UserHistoryOperation::max, UserHistoryOperation::min};
+  for (auto &op : ops) {
+    results[op] = {};
+    labels[op] = {};
+  }
+
+  // If "packages" not provided, output history for all packages
+  auto &requested_packages = output_params.packages;
+  Dictionary<std::shared_ptr<StateDescriptor>> packages;
+  if (requested_packages.empty()) {
+    packages = pm->packages.AllPackages();
+  } else {
+    for (const auto &pkg : pm->packages.AllPackages()) {
+      if (std::find(requested_packages.begin(), requested_packages.end(), pkg.first) !=
+          requested_packages.end()) {
+        packages[pkg.first] = pkg.second;
+      }
+    }
```

Comment:
This seems a little backwards. In the limit of `n_total_packages > n_requested_packages`, this is probably better:
```suggestion
    const auto &all_packages = pm->packages.AllPackages();
    for (const auto & pkg_name : requested_packages) {
      const auto &it = all_packages.find(pkg_name);
      if (it != all_packages.end()) {
        packages[it.first] = it.second;
      }
   }
```

Reply:
Yeah this is an improvement, fixed, thanks